File: backend/app/schemas/project.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import date, datetime


# 项目基础模型
class ProjectBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=200, description="项目名称")
    describe: Optional[str] = Field(None, description="项目描述")
    leader: str = Field(..., min_length=1, max_length=100, description="项目负责人")
    leader_id: Optional[int] = Field(None, description="负责人ID")
    status: str = Field(..., min_length=1, max_length=50, description="项目状态")
    start_date: Optional[date] = Field(None, description="项目开始日期")
    end_date: Optional[date] = Field(None, description="项目结束日期")


# 创建项目请求模型
class ProjectCreate(ProjectBase):
    contract_date: Optional[date] = Field(None, description="合同日期")
    contract_no: Optional[str] = Field(None, max_length=100, description="合同编号")
    tax_rate: Optional[float] = Field(0.13, ge=0, description="税率")
    contract_amount: Optional[float] = Field(0.0, ge=0, description="合同金额")
    revenue: Optional[float] = Field(0.0, ge=0, description="收入")
    target_profit: Optional[float] = Field(0.0, ge=0, description="目标利润")
    actual_profit: Optional[float] = Field(0.0, ge=0, description="实际利润")
    created_by_id: Optional[int] = Field(None, description="创建人ID")
    # 成本相关
    budget_total_cost: Optional[float] = Field(0.0, ge=0, description="预算总成本")
    actual_total_cost: Optional[float] = Field(0.0, ge=0, description="实际总成本")
    material_budget: Optional[float] = Field(0.0, ge=0, description="物料预算")
    material_actual: Optional[float] = Field(0.0, ge=0, description="实际物料成本")
    outsourcing_budget: Optional[float] = Field(0.0, ge=0, description="外包预算")
    outsourcing_actual: Optional[float] = Field(0.0, ge=0, description="实际外包成本")
    indirect_budget: Optional[float] = Field(0.0, ge=0, description="间接预算")
    indirect_actual: Optional[float] = Field(0.0, ge=0, description="实际间接成本")
    labor_budget: Optional[float] = Field(0.0, ge=0, description="人力预算")
    labor_actual: Optional[float] = Field(0.0, ge=0, description="实际人力成本")


# 更新项目请求模型
class ProjectUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200, description="项目名称")
    leader: Optional[str] = Field(None, min_length=1, max_length=100, description="项目负责人")
    leader_id: Optional[int] = Field(None, description="负责人ID")
    status: Optional[str] = Field(None, min_length=1, max_length=50, description="项目状态")
    progress: Optional[int] = Field(None, ge=0, le=100, description="项目进度")
    start_date: Optional[date] = Field(None, description="项目开始日期")
    end_date: Optional[date] = Field(None, description="项目结束日期")
    contract_date: Optional[date] = Field(None, description="合同日期")
    contract_no: Optional[str] = Field(None, max_length=100, description="合同编号")
    tax_rate: Optional[float] = Field(None, ge=0, description="税率")
    contract_amount: Optional[float] = Field(None, ge=0, description="合同金额")
    revenue: Optional[float] = Field(None, ge=0, description="收入")
    target_profit: Optional[float] = Field(None, ge=0, description="目标利润")
    actual_profit: Optional[float] = Field(None, ge=0, description="实际利润")
    # 成本相关
    budget_total_cost: Optional[float] = Field(None, ge=0, description="预算总成本")
    actual_total_cost: Optional[float] = Field(None, ge=0, description="实际总成本")
    material_budget: Optional[float] = Field(None, ge=0, description="物料预算")
    material_actual: Optional[float] = Field(None, ge=0, description="实际物料成本")
    outsourcing_budget: Optional[float] = Field(None, ge=0, description="外包预算")
    outsourcing_actual: Optional[float] = Field(None, ge=0, description="实际外包成本")
    indirect_budget: Optional[float] = Field(None, ge=0, description="间接预算")
    indirect_actual: Optional[float] = Field(None, ge=0, description="实际间接成本")
    labor_budget: Optional[float] = Field(None, ge=0, description="人力预算")
    labor_actual: Optional[float] = Field(None, ge=0, description="实际人力成本")


# 项目响应模型
class ProjectResponse(ProjectBase):
    id: int = Field(..., description="项目ID")
    progress: int = Field(..., ge=0, le=100, description="项目进度")
    contract_date: Optional[date] = Field(None, description="合同日期")
    contract_no: Optional[str] = Field(None, description="合同编号")
    tax_rate: float = Field(..., ge=0, description="税率")
    contract_amount: float = Field(..., ge=0, description="合同金额")
    revenue: float = Field(..., ge=0, description="收入")
    target_profit: float = Field(..., ge=0, description="目标利润")
    actual_profit: float = Field(..., ge=0, description="实际利润")
    created_by_id: Optional[int] = Field(None, description="创建人ID")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    is_deleted: bool = Field(..., description="是否删除")
    # 成本相关
    budget_total_cost: float = Field(..., ge=0, description="预算总成本")
    actual_total_cost: float = Field(..., ge=0, description="实际总成本")
    material_budget: float = Field(..., ge=0, description="物料预算")
    material_actual: float = Field(..., ge=0, description="实际物料成本")
    outsourcing_budget: float = Field(..., ge=0, description="外包预算")
    outsourcing_actual: float = Field(..., ge=0, description="实际外包成本")
    indirect_budget: float = Field(..., ge=0, description="间接预算")
    indirect_actual: float = Field(..., ge=0, description="实际间接成本")
    labor_budget: float = Field(..., ge=0, description="人力预算")
    labor_actual: float = Field(..., ge=0, description="实际人力成本")
    
    class Config:
        from_attributes = True


# 任务基础模型
class TaskBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100, description="任务名称")
    description: Optional[str] = Field(None, max_length=500, description="任务描述")
    start_date: date = Field(..., description="任务开始日期")
    end_date: date = Field(..., description="任务结束日期")
    status: str = Field(..., min_length=1, max_length=20, description="任务状态")
    progress: float = Field(..., ge=0, le=100, description="任务进度")
    assigned_to: Optional[str] = Field(None, max_length=50, description="任务负责人")


# 创建任务请求模型
class TaskCreate(TaskBase):
    pass


# 更新任务请求模型
class TaskUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="任务名称")
    description: Optional[str] = Field(None, max_length=500, description="任务描述")
    start_date: Optional[date] = Field(None, description="任务开始日期")
    end_date: Optional[date] = Field(None, description="任务结束日期")
    status: Optional[str] = Field(None, min_length=1, max_length=20, description="任务状态")
    progress: Optional[float] = Field(None, ge=0, le=100, description="任务进度")
    assigned_to: Optional[str] = Field(None, max_length=50, description="任务负责人")


# 任务响应模型
class TaskResponse(TaskBase):
    id: int = Field(..., description="任务ID")
    project_id: int = Field(..., description="项目ID")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    
    class Config:
        from_attributes = True

File: frontend/src/api/project.ts
import request from './axios'

// 项目相关API

export interface Project {
  id: number
  name: string
  describe: string
  start_date: string
  "计划结束日期": string
  "实际结束日期"?: string
  budget_total_cost: number
  actual_total_cost: number
  contract_amount: number
  status: string
  progress: number
  leader: string
  created_at: string
  updated_at: string
}

export interface ProjectCreate {
    name: string
    describe: string
    start_date: string | null
    end_date: string | null
    status: string
    leader: string
    // 合同信息
    contract_no: string
    contract_date: string | null
    contract_amount: number
    tax_rate: number
    revenue: number
    target_profit: number
}

export interface ProjectUpdate {
    name?: string
    describe?: string
    start_date?: string | null
    end_date?: string | null
    status?: string
    progress?: number
    leader?: string
    // 合同信息
    contract_no?: string
    contract_date?: string | null
    contract_amount?: number
    tax_rate?: number
    revenue?: number
    target_profit?: number
}

// 任务相关接口
export interface Task {
  task_id: number
  project_id: number
  project_name?: string
  task_name: string
  start_date: string
  end_date: string
  actual_end_date?: string
  difdays?: number
  completed?: boolean
  progress?: number
  assignee?: string
  create_time?: string
  update_time?: string
}

export interface TaskCreate {
  task_name: string
  start_date: string
  end_date: string
  assignee?: string
  progress?: number
}

/**
 * 获取项目列表
 * @param params 查询参数
 */
export const getProjects = (params?: { page?: number; size?: number; status?: string; name?: string }) => {
  return request.get<{ items: Project[]; total: number; page: number; size: number; total_pages: number }>('v1/projects/', { params })
}

/**
 * 检查项目名称是否已存在
 * @param name 项目名称
 * @param excludeId 排除的项目ID（用于编辑模式）
 */
export const checkProjectName = (name: string, excludeId?: number) => {
  return request.get<{ exists: boolean; message: string }>(`v1/projects/check-name/${name}`, {
    params: excludeId ? { exclude_id: excludeId } : {},
    cache: false // 移除缓存，确保每次都获取最新结果
  })
}

/**
 * 获取项目详情
 * @param projectId 项目ID
 */
export const getProjectDetail = (projectId: number) => {
  return request.get<Project>(`v1/projects/${projectId}/`, {
    cache: true // 添加缓存
  })
}

/**
 * 创建项目
 * @param project 项目信息
 */
export const createProject = (project: ProjectCreate) => {
  return request.post<{ id: number }>('v1/projects/', project)
}

/**
 * 更新项目
 * @param projectId 项目ID
 * @param project 项目信息
 */
export const updateProject = (projectId: number, project: ProjectUpdate) => {
  return request.put<Project>(`v1/projects/${projectId}/`, project)
}

/**
 * 删除项目
 * @param projectId 项目ID
 */
export const deleteProject = (projectId: number) => {
  return request.delete(`v1/projects/${projectId}/`)
}

/**
 * 获取项目任务列表
 * @param projectId 项目ID
 */
export const getProjectTasks = (projectId: number) => {
  return request.get<Task[]>(`v1/projects/${projectId}/tasks/`, {
    cache: false // 禁用缓存，确保获取最新数据
  })
}

/**
 * 获取所有项目的任务列表（用于看板）
 */
export const getAllTasks = () => {
  return request.get<Task[]>('v1/projects/tasks', {
    cache: false // 禁用缓存，确保获取最新数据
  })
}

/**
 * 创建项目任务
 * @param projectId 项目ID
 * @param task 任务信息
 */
export const createProjectTask = (projectId: number, task: TaskCreate) => {
  return request.post<Task>(`v1/projects/${projectId}/tasks/`, task)
}

/**
 * 获取项目成本汇总
 * @param projectId 项目ID
 */
export const getProjectCosts = (projectId: number) => {
  return request.get(`v1/projects/${projectId}/costs/`, {
    cache: false // 禁用缓存，确保获取最新数据
  })
}

/**
 * 获取项目概览统计
 */
export const getProjectOverviewStatistics = () => {
  return request.get<{ total_projects: number; status_counts: Record<string, number>; total_contract_amount: number; total_budget_cost: number; total_actual_cost: number; cost_variance: number; avg_progress: number }>(`v1/projects/statistics/overview`, {
    cache: true // 添加缓存
  })
}

/**
 * 获取外包服务类型列表
 */
export const getOutsourcingServiceTypes = () => {
  return request.get(`v1/resource/outsourcing-service-types/`, {
    cache: true // 添加缓存
  })
}

/**
 * 获取间接成本类型列表
 */
export const getIndirectCostTypes = () => {
  return request.get(`v1/resource/indirect-cost-types/`, {
    cache: true // 添加缓存
  })
}

/**
 * 批量创建项目任务
 * @param tasks 任务列表
 */
export const batchCreateProjectTasks = (tasks: { project_id: number; task_name: string; start_date: string; end_date: string; assignee?: string; progress?: number }[]) => {
  return request.post(`v1/projects/tasks/batch`, tasks)
}

/**
 * 批量上传任务附件
 * @param attachments 附件列表
 */
export const batchUploadTaskAttachments = (attachments: { task_id: number; file_name: string; file_url: string; file_size?: number }[]) => {
  return request.post(`v1/projects/task-attachments/batch`, attachments)
}

/**
 * 批量创建材料成本
 * @param costs 成本列表
 */
export const batchCreateMaterialCosts = (costs: { project_id: number; material_id: string; quantity: number; unit_price: number; total_cost: number; cost_date: string }[]) => {
  return request.post(`v1/cost/material/batch`, costs)
}

/**
 * 批量创建人工成本
 * @param costs 成本列表
 */
export const batchCreateLaborCosts = (costs: { project_id: number; employee_id: string; work_date: string; hours: number; hourly_rate: number; total_cost: number }[]) => {
  return request.post(`v1/cost/labor/batch`, costs)
}

/**
 * 批量创建外包成本
 * @param costs 成本列表
 */
export const batchCreateOutsourcingCosts = (costs: { project_id: number; service_type_id: number; cost_amount: number; cost_date: string; description?: string }[]) => {
  return request.post(`v1/cost/outsourcing/batch`, costs)
}

/**
 * 批量创建间接成本
 * @param costs 成本列表
 */
export const batchCreateIndirectCosts = (costs: { project_id: number; cost_type_id: number; cost_amount: number; cost_date: string; description?: string }[]) => {
  return request.post(`v1/cost/indirect/batch`, costs)
}

/**
 * 获取任务附件列表
 * @param projectId 项目ID
 * @param taskId 任务ID
 */
export const getTaskAttachments = (projectId: number, taskId: number) => {
  return request.get(`v1/projects/${projectId}/task-attachments/${taskId}`, {
    cache: false // 禁用缓存，确保获取最新数据
  })
}

/**
 * 获取项目文档列表
 * @param projectId 项目ID
 */
export const getProjectDocuments = (projectId: number) => {
  return request.get(`v1/projects/${projectId}/documents/`, {
    cache: false // 禁用缓存，确保获取最新数据
  })
}

/**
 * 批量创建项目任务附件
 * @param attachments 附件列表
 */
export const batchCreateProjectTaskAttachments = (attachments: { task_id: number; file_name: string; file_url: string; file_size?: number }[]) => {
  return request.post(`v1/projects/task-attachments/batch`, attachments)
}

/**
 * 获取项目任务附件列表
 * @param projectId 项目ID
 * @param taskId 任务ID
 */
export const getProjectTaskAttachments = (projectId: number, taskId: number) => {
  return request.get(`v1/projects/${projectId}/task-attachments/${taskId}`, {
    cache: false // 禁用缓存，确保获取最新数据
  })
}